Fixed Camera Stream trait to match the latest Google API.

wir3z opened the pull request:
Fixed Camera Stream trait to match the latest Google API.  Moved the stream protocol selection to the driver level to allow for combinates of stream sources.  NOTE:  webRTC protocol support has not been completed.

wir3z commented:
The updates to Google Camera Trait broke the current implementation.

mbudnek commented on google-home-community.groovy:
```diff
@@ -2520,10 +2533,8 @@ private deviceStateForTrait_Brightness(deviceTrait, device) {
-        cameraStreamReceiverAppId:    null,
-        cameraStreamAuthToken:        null,
-        cameraStreamProtocol:         deviceTrait.cameraStreamSupportedProtocols
+        cameraStreamAccessUrl:        device.currentValue(deviceTrait.cameraStreamURLAttribute),
+        cameraStreamProtocol:         device.currentValue(deviceTrait.cameraStreamProtocolAttribute),
```
Are these needed?  The documentation doesn't list any states for the CameraStream trait.

wir3z replied:
You are correct, not needed so they are now removed.

mbudnek reviewed (commented):
Sorry I let this PR languish for so long.  It looks good, but it would be nice to maintain backwards-compatibility with existing device trait configurations.  As-is, existing users will have to re-configure their CameraStream traits before they'll work with this change.

wir3z commented on google-home-community.groovy:
```diff
@@ -432,27 +434,27 @@ private deviceTraitPreferences_Brightness(deviceTrait) {
     section("Stream Camera") {
         input(
-            name: "${deviceTrait.name}.cameraStreamAttribute",
+            name: "${deviceTrait.name}.cameraStreamURLAttribute",
```
Previous implementation only permitted a single camera protocol since it was based in the app.  Changes moved this selection to the driver to allow a multitude of camera protocol types which is breaking backwards compatibility in this case.

wir3z replied:
Need to break backward compatibility to move the URLs to the driver level.

wir3z commented on google-home-community.groovy:
```diff
@@ -432,27 +434,27 @@ private deviceTraitPreferences_Brightness(deviceTrait) {
-            options: googleCameraStreamSupportedProtocols,
-            multiple: true,
-            required: true,
+            name: "${deviceTrait.name}.cameraStreamProtocolAttribute",
```
Previous implementation only permitted a single camera protocol since it was based in the app.  Changes moved this selection to the driver to allow a multitude of camera protocol types which is breaking backwards compatibility in this case.

wir3z replied:
Impacts the above concern as well

wir3z replied:
User would need to select the camera protocol from the driver.

wir3z replied:
Need to break backward compatibility to move the URLs to the driver level.

mbudnek commented on google-home-community.groovy:
```diff
@@ -1928,7 +1930,18 @@ private executeCommand_BrightnessAbsolute(deviceInfo, command) {
-    return [[:], [:]]
+    checkMfa(deviceInfo.deviceType, "Display", command)
+    def cameraStreamTrait = deviceInfo.deviceType.traits.CameraStream
+    deviceInfo.device."${cameraStreamTrait.cameraStreamCommand}"()
```
Should this command get passed the `command.params.SupportedStreamProtocols` so the device can select a protocol from the list of protocols that the target device supports?

wir3z replied:
Are you thinking on the driver side you select everything the driver device can support, and then only link through the protocol the display device supports during the execute?  

Right now, if you pass a format to google that it cannot display, it just gives you a black screen with "smart home camera" on it (so it gracefully handles the negative case).  I didn't make it more complex since you don't know what the display can handle until you actually call up the camera to display the stream.

Thoughts?

mbudnek replied:
The list of protocols that Google sends in `command.params.SupportedStreamProtocols` is the list of protocols that the target display supports.  If we pass that to the camera driver then it can pick one of the protocols that both it and the display support.

wir3z replied:
Working on this.  Will need to add URL input fields for all the supported formats at the driver level.

wir3z replied:
Fixed.  Multiple URL's can be specified in the generic driver which get matched to the display device protocols.

mbudnek commented on google-home-community.groovy:
```diff
@@ -2886,9 +2894,8 @@ private attributesForTrait_Brightness(deviceTrait) {
 private attributesForTrait_CameraStream(deviceTrait) {
     return [
-        cameraStreamSupportedProtocols: deviceTrait.cameraStreamSupportedProtocols,
+        cameraStreamSupportedProtocols: ["progressive_mp4", "hls", "dash", "smooth_stream"],
```
Should we get this list from a device attribute instead of assuming all devices support all streaming protocols?

wir3z replied:
Issue I'm seeing is we don't know what each device's supported protocol is.  I suppose maybe I could query all of them, and then generate a list to pass back?  If the display device supports Chromecast, then it should support all of these.

mbudnek replied:
I'm pretty sure this attribute tells Google what protocols the _source_ supports, not the display.  I was thinking something like what Hubitat does with `supportedThermostatModes`.  Add an attribute to the driver that can be used to report supported streaming protocols to the app.  For the generic driver the user would have to set it, but for any specific devices it could be hard-coded into the driver.

wir3z replied:
Issue is what the app supports not the driver.  It's only called at the start of the sync process, and that result is then loaded into all the camera devices reported back to Google.   The only time the driver level support is pulled is during the execute via deviceInfo.device.currentValue(deviceInfo.deviceType.traits.CameraStream.cameraStreamProtocolAttribute.  

It used to be like supportedThermostatModes where you could pick the supported protocols, but that prevented the user from linking more than one.  I can add it back to the app, but then it becomes confusing for the user since they need to configure it in the app as well as the driver level.  Thoughts?

wir3z replied:
> Issue is what the app supports not the driver. It's only called at the start of the sync process, and that result is then loaded into all the camera devices reported back to Google. The only time the driver level support is pulled is during the execute via deviceInfo.device.currentValue(deviceInfo.deviceType.traits.CameraStream.cameraStreamProtocolAttribute.
> 
> It used to be like supportedThermostatModes where you could pick the supported protocols, but that prevented the user from linking more than one. I can add it back to the app, but then it becomes confusing for the user since they need to configure it in the app as well as the driver level. Thoughts?

I re-looked at this, but still running back to this issue.  The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?

mbudnek replied:
> I re-looked at this, but still running back to this issue. The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?

I see what you mean now.  I think it would make sense to move the loop line 2910 down into the following loop and pass both the device and the trait into the `attributesForTrait_*` methods.  That way the initial SYNC would have access to the devices themselves to pull things like supported protocols.

The other option would be to move it back to being a trait level setting and make users define multiple device types for devices that support different protocols.  That would also make sense to me, since all instances of one type of camera will always support the same protocols.

wir3z replied:
Let me take a look again with those suggestions.  Thanks!

wir3z replied:
> > I re-looked at this, but still running back to this issue. The framework isn't structured to pull parameters from the driver during the sync, unless I'm missing something?
> 
> I see what you mean now. I think it would make sense to move the loop line 2910 down into the following loop and pass both the device and the trait into the `attributesForTrait_*` methods. That way the initial SYNC would have access to the devices themselves to pull things like supported protocols.
> 
> The other option would be to move it back to being a trait level setting and make users define multiple device types for devices that support different protocols. That would also make sense to me, since all instances of one type of camera will always support the same protocols.

Struggling a bit on this one.  Won't this change break all the existing code actions since the traits are global, and not device specific?  I understand your concern with the protocols, but from the testing I've done, if Google asks for something the driver cannot handle, it handles it gracefully on their end.

mbudnek replied:
The SYNC request handler already does device-specific stuff to get default rooms, there's no reason it can't do other device-specific lookups.  I'm basically just suggesting to change it from this:

```
    (deviceTypes() + [modeSceneDeviceType()]).each { deviceType ->
        def traits = deviceType.traits.collect { traitType, deviceTrait ->
            "action.devices.traits.${traitType}"
        }
        def attributes = [:]
        deviceType.traits.each { traitType, deviceTrait ->
            attributes += "attributesForTrait_${traitType}"(deviceTrait)
        }
        deviceType.devices.each { device ->
            //...
        }
```

To something like this:

```
    (deviceTypes() + [modeSceneDeviceType()]).each { deviceType ->
        def traits = deviceType.traits.collect { traitType, deviceTrait ->
            "action.devices.traits.${traitType}"
        }
        deviceType.devices.each { device ->
            def attributes = [:]
            deviceType.traits.each { traitType, deviceTrait ->
                attributes += "attributesForTrait_${traitType}"(deviceTrait, device)
            }

            // ...
        }
```

The `attributesForTrait_*` methods would need to accept the second parameter, but most could just ignore it and always return the same thing for all devices.

mbudnek reviewed (commented):
OK, so assuming breaking compatibility is OK, then I have a couple of other comments below.

wir3z commented:
Added the arm/disarm trait to this commit.  Can you do a  sanity check on the CheckMFA in the arm/disarm execute command?

nipuna-perera commented:
Hey guys, what needs to be done in order to get this merged? Would love to help if possible.

mbudnek commented:
Sorry, I definitely let this languish for too long.  As for getting it merged, it looks mostly OK.  I would like to understand the changes to `checkMfa` a bit more first.  What kind of devices does the whole `codePosition` thing exist to support?  Looking at Hubitat's documentation, the SecurityKeypad capability doesn't seem to accept a parameter to its arm/disarm commands.  I would also like to see the `cameraStreamSupportedProtocols` get supplied by the driver.  As-is, I think it's very possible Google will try to select a protocol that the camera doesn't actually support.

wir3z commented:
> Sorry, I definitely let this languish for too long. As for getting it merged, it looks mostly OK. I would like to understand the changes to `checkMfa` a bit more first. What kind of devices does the whole `codePosition` thing exist to support? Looking at Hubitat's documentation, the SecurityKeypad capability doesn't seem to accept a parameter to its arm/disarm commands. I would also like to see the `cameraStreamSupportedProtocols` get supplied by the driver. As-is, I think it's very possible Google will try to select a protocol that the camera doesn't actually support.

The idea behind the "codePosition" was to allow for pin codes to come from the driver level vs GHC local/global.  For the cases of alarm panels, the panel owns the pin code so it should come from the driver space.  Currently it's the alarm trait, but a lock trait or any other trait that requires a challenge pin could use it.

Making the driver supply "cameraStreamSupportedProtocols" would be easy.  The driver already knows which sources have been specified, so I could pull that back into GHC.

I do have a "dishwasher" and "sensor" traits that I am waiting to make a pull request on pending getting these cleaned up.

mbudnek commented:
> The idea behind the "codePosition" was to allow for pin codes to come from the driver level vs GHC local/global. For the cases of alarm panels, the panel owns the pin code so it should come from the driver space. Currently it's the alarm trait, but a lock trait or any other trait that requires a challenge pin could use it.

Right, I get that's the idea behind the `useDevicePinCodes` setting.  But why are you returning the index of the matched PIN and passing that back to the device's arm command?  Looking at Hubitat's device capabilities, none of the stock ones accept a parameter to any of their arm commands.

> I do have a "dishwasher" and "sensor" traits that I am waiting to make a pull request on pending getting these cleaned up.

Awesome!  It would be easier to review them if you could separate each them into separate branches, but if not it's not the end of the world.

wir3z commented:
> > The idea behind the "codePosition" was to allow for pin codes to come from the driver level vs GHC local/global. For the cases of alarm panels, the panel owns the pin code so it should come from the driver space. Currently it's the alarm trait, but a lock trait or any other trait that requires a challenge pin could use it.
> 
> Right, I get that's the idea behind the `useDevicePinCodes` setting. But why are you returning the index of the matched PIN from `checkMfa` and passing that back to the device's arm command? Looking at Hubitat's device capabilities, none of the stock ones accept a parameter to any of their arm commands.
> 

Hubitat has failed to document that the arm/disarm have a parameter to indicate the source of those operations.  If you don't handle it in the driver call, then you get an exception.

> > I do have a "dishwasher" and "sensor" traits that I am waiting to make a pull request on pending getting these cleaned up.
> 
> Awesome! It would be easier to review them if you could separate each them into separate branches, but if not it's not the end of the world.

Will do, sorry, I blended those last 2 traits into the same pull request which is just bad commit hygiene.

mbudnek commented on README.md:
```diff
@@ -121,16 +152,18 @@ The Brightness trait is primarily used for devices like dimmer switches and ligh
 - Current Brightness Attribute: The device attribute used to query the current brightness level of the device.  Should be in the range 0-100.  Maps to the `level` attribute by default.
 - Set Brightness Command: A device command used to set the brightness of the device.  Should accept a brightness level in the range 0-100.  Maps to the `setLevel` command by default.
-
+	
```
Whitespace

wir3z replied:
Fixed on next commit.

mbudnek commented on google-home-community.groovy:
```diff
@@ -411,6 +430,142 @@ def deviceTraitDelete(deviceTrait) {
+            required: true,
+            submitOnChange: true
+        )
+        if (deviceTrait.useDevicePinCodes == true) {
```
Could this be configured at the device type level?  It seems like this should be a selectable option for any trait.

wir3z replied:
Good point.  looking at this.

wir3z replied:
Right now this is at the trait level.  You would prefer it up one level at the device level?  When the user is adding traits to a device?

mbudnek replied:
Yeah, it doesn't seem like it should trait-specific to me.  Any device could offer device-specific PIN codes in theory, and I think it would make more sense as a peer of the existing PIN code settings at the device type level.

wir3z replied:
I agree.  The initial user story I was thinking of was if a device had a lock + armDisarm trait, then they could have 2 sets of device level pins.  Thinking about it, that's probably a highly unlikely scenario.

I'll move it up a level with a toggle to use them from the device driver context, or the device context in GHC.

wir3z replied:
Moved this to the device level from the trait level.  There is a boolean slider to allow the matched index to be send back as a command parameter:
                        input(
                            name: "${devicePropertyName}.returnUserIndexToDevice",
                            title: "Select to return the user index with the device command on a pincode match. " +
                                    "Not all device drivers support this operation.",
                            type: "bool",
                            defaultValue: false,
                            submitOnChange: true
                        )

This is available on all devices, but I only implemented it on the arm trait and lock trait.  What are your thoughts on this?  I can add it to all if think it's required,  Just becomes a bit ore complicated on devices that already pass a parameter back (such as brightness).

wir3z replied:
Moved this setting to the arm and lock traits only.

mbudnek commented:
> Hubitat has failed to document that the arm/disarm have a parameter to indicate the source of those operations. If you don't handle it in the driver call, then you get an exception.

Hmm, if there are such drivers that's fine, but I would really like to get rid of the `noMatchValue`.  I don't think it makes sense as-is anyway.

wir3z commented on google-home-community.groovy:
```diff
@@ -1830,18 +2013,44 @@ private handleExecuteRequest(request) {
+
+@SuppressWarnings(['InvertedIfElse', 'NestedBlockDepth'])
+private checkMfa(deviceInfo, commandType, command, noMatchValue) {
+    def matchPosition = noMatchValue
```
I see what you mean.  Let me rework this portion.

wir3z replied:
Removed this.

wir3z commented on google-home-community.groovy:
```diff
@@ -1850,16 +2059,34 @@ private checkMfa(deviceType, commandType, command) {
+            def positionMatchTrait = checkTraitPinMatch(deviceInfo, command, noMatchValue)
+
+            // return pincode matches in the order of trait -> device -> global -> noMatchValue
+            if (positionMatchTrait != noMatchValue) {
```
Same as above, will rework it.

wir3z replied:
Changing this to null, and removing all the "-1" from CheckMfa

wir3z commented:
Addressed the supported protocols and fixed an issue with the nested "sections" that was crashing the latest Hubitat version.  Working on the pin match.

mbudnek commented on google-home-community.groovy:
```diff
@@ -386,9 +407,7 @@ def deviceTraitPreferences(deviceTrait) {
         title: "Preferences For ${GOOGLE_DEVICE_TRAITS[deviceTrait.type]} Trait",
         nextPage: "deviceTypePreferences"
     ) {
-        section {
-            "deviceTraitPreferences_${deviceTrait.type}"(deviceTrait)
-        }
```
Good timing.  I was just trying to figure this error out after some people reported it in the forum thread.  I went ahead and pulled this change out to a separate commit (c3e51e1973c951477ed6da8a824d3aa839ecb51e) so it wouldn't be broken while you work on this PR.  Thanks for catching this; I was certainly confused.

mbudnek commented on google-home-community.groovy:
```diff
@@ -1830,18 +2039,44 @@ private handleExecuteRequest(request) {
+            }
+        }
+    }
+    return(matchPosition)
```
```suggestion
    return matchPosition
```

mbudnek commented on google-home-community.groovy:
```diff
@@ -1850,16 +2085,34 @@ private checkMfa(deviceType, commandType, command) {
         }
     }
+
+    return(matchPosition)
```
```suggestion
    return matchPosition
```

mbudnek commented on google-home-community.groovy:
```diff
@@ -1904,9 +2157,57 @@ private executeCommand_ActivateScene(deviceInfo, command) {
+private executeCommand_ArmDisarm(deviceInfo, command) {
+    def armDisarmTrait = deviceInfo.deviceType.traits.ArmDisarm
+    def checkValue = "${armDisarmTrait.armValues["disarmed"]}"
+    def codePosition
```
This doesn't need to be declared outside the `if` block.

wir3z replied:
They are used inside the if and the else though?

mbudnek replied:
Sure, but they're never used outside the scope where they're set.  Just declare this variable in each branch.

wir3z replied:
I see what you mean.  Moved codePosition inside each block.  checkValue needs a default that isn't set in 2/3 of those if-else blocks.

mbudnek commented on google-home-community.groovy:
```diff
@@ -1904,9 +2157,57 @@ private executeCommand_ActivateScene(deviceInfo, command) {
+        } else {
+            deviceInfo.device."${armDisarmTrait.cancelCommand}"()
+        }
+    } else {
+        // if Google returns arm=false, that indicates disarm, otherwise execute the matching alarm level command
+        if (command.params.arm == false) {
```
Does this need to be nested inside the `else` instead of an `if ... else if ... else`?

wir3z replied:
The arm cancel trumps the arm commands (they are separate).  The arm command is disarm, level 0, level 1....

If cancel was part of the arm, yes, I'm just not seeing how I can structure it without the nest if else.

mbudnek replied:
Aren't they logically equivilent?  `if ... else { if ... else ... }` should have the same truth table as `if ... else if ... else ...`:

| cancel | arm | nested | not nested
| ------- |------ | -------- | ----------
| true     | true | if         | if
| true     | false | if        | if
| false    | true  | else { if } | else if
| false    | false | else { else } | else

mbudnek replied:
In fact, I would probably structure this whole thing as
```
if (command.params.cancel) {
    // cancel logic
} else if (command.params.arm) {
    // arm logic
} else {
    // disarm logic
}
```

wir3z replied:
Agreed, that's cleaner.  Done.

mbudnek reviewed (commented):
Few more little nitpicks, but this is looking really good

mbudnek commented on google-home-community.groovy:
```diff
@@ -4355,15 +4306,19 @@ private deviceTypeTraitsFromSettings(deviceTypeName) {
+        confirmCommands:          [],
+        secureCommands:           [],
+        pinCodes:                 [],
+        useDevicePinCodes:        [],
+        pinCodeAttribute:         [],
+        pinCodeValue:             [],
+        returnUserIndexToDevice:  [],
```
An empty list doesn't make much sense as the default value for these, perhaps something like this:
```suggestion
        useDevicePinCodes:        false,
        pinCodeAttribute:         null,
        pinCodeValue:             null,
        returnUserIndexToDevice:  false,
```

wir3z replied:
Thanks, and done.

mbudnek commented on google-home-community.groovy:
```diff
@@ -2122,6 +2090,15 @@ private controlScene(options) {
     device."${options.command}"()
 }
 
+@SuppressWarnings('UnusedPrivateMethod')
```
This method is used directly, you shouldn't need to suppress an unused warning.

wir3z replied:
Done!

mbudnek commented on google-home-community.groovy:
```diff
@@ -2122,6 +2090,15 @@ private controlScene(options) {
 }
 
+@SuppressWarnings('UnusedPrivateMethod')
+private issueCommand(deviceInfo, command, codePosition) {
```
```suggestion
private issueCommandWithCodePosition(deviceInfo, command, codePosition) {
```

wir3z replied:
Done!

mbudnek commented on google-home-community.groovy:
```diff
@@ -328,12 +328,47 @@ def deviceTypePreferences(deviceType) {
+                            type: "bool",
+                            defaultValue: false,
+                            submitOnChange: true
+                        )
```
Does this setting make sense at the device type level?  Only Lock/Unlock and Arm/Disarm support it.

wir3z replied:
I've moved it back to the trait level for those two traits.

mbudnek commented on google-home-community.groovy:
```diff
@@ -59,6 +59,17 @@
+//                 - Added Arm/Disarm Trait
+//                 - Added ability for the app to use trait level pin codes retrieved from the device for
+//                   Arm/Disarm and Lock/Unlock
+//                 - Modified checkMfa to accept deviceInfo (instead of deviceType)
+//                   Added support for trait level pincodes, as well as returning the matching user position or
+//                   no match error value on pincode challenge in the order of trait -> device -> global -> null
```
Nitpick: I would drop the bit about modifying `checkMfa` since it's not really a user-visible change.

wir3z replied:
Agreed.
